File: src/js/src/plugins.ts
import { readdirSync, realpathSync, statSync } from "node:fs";
import { delimiter, isAbsolute, join } from "node:path";

export function discover(): Map<string, string> {
  const plugins = new Map<string, string>();
  for (const directory of (process.env.PATH ?? "").split(delimiter)) {
    if (!isAbsolute(directory)) continue;
    let entries: string[];
    try {
      entries = readdirSync(directory);
    } catch {
      continue;
    }
    for (const entry of entries) {
      let filename = entry;
      if (process.platform === "win32") {
        if (!filename.endsWith(".exe")) continue;
        filename = filename.slice(0, -4);
      }
      if (!/^maw-[a-z][a-z0-9-]*$/.test(filename)) continue;
      const name = filename.slice(4);
      if (plugins.has(name)) continue;
      try {
        const path = realpathSync(join(directory, entry));
        const info = statSync(path);
        if (!info.isFile()) continue;
        if (process.platform !== "win32" && (info.mode & 0o111) === 0) continue;
        plugins.set(name, path);
      } catch {
        // Unreadable files and broken symlinks are not eligible plugins.
      }
    }
  }
  return plugins;
}

export async function execute(path: string, args: string[]): Promise<number> {
  try {
    const child = Bun.spawn([path, ...args], {
      stdin: "inherit",
      stdout: "inherit",
      stderr: "inherit",
      env: process.env,
    });
    let interrupted = false;
    const interrupt = () => {
      interrupted = true;
      child.kill("SIGKILL");
    };
    process.on("SIGINT", interrupt);
    try {
      const code = await child.exited;
      return interrupted ? 1 : child.signalCode ? 126 : code;
    } finally {
      process.off("SIGINT", interrupt);
    }
  } catch (error) {
    console.error(`maw: cannot execute ${path}: ${error instanceof Error ? error.message : error}`);
    return 126;
  }
}

File: src/rs/src/plugins.rs
use std::collections::BTreeMap;
use std::ffi::OsString;
use std::path::PathBuf;
use std::process::Command;

pub fn discover() -> BTreeMap<String, PathBuf> {
    let mut plugins = BTreeMap::new();
    let path = std::env::var_os("PATH").unwrap_or_default();
    for dir in std::env::split_paths(&path).filter(|dir| dir.is_absolute()) {
        let entries = match std::fs::read_dir(dir) {
            Ok(entries) => entries,
            Err(_) => continue,
        };
        for entry in entries.flatten() {
            let filename = entry.file_name();
            let filename = match filename.to_str() {
                Some(name) => name,
                None => continue,
            };
            #[cfg(windows)]
            let filename = match filename.strip_suffix(".exe") {
                Some(name) => name,
                None => continue,
            };
            let name = match filename.strip_prefix("maw-") {
                Some(name) => name,
                None => continue,
            };
            if !valid_name(name)
                || matches!(name, "go" | "rs" | "js" | "zig")
                || plugins.contains_key(name)
            {
                continue;
            }
            let path = match std::fs::canonicalize(entry.path()) {
                Ok(path) => path,
                Err(_) => continue,
            };
            let metadata = match path.metadata() {
                Ok(metadata) if metadata.is_file() => metadata,
                _ => continue,
            };
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                if metadata.permissions().mode() & 0o111 == 0 {
                    continue;
                }
            }
            #[cfg(not(unix))]
            let _ = metadata;
            plugins.insert(name.to_owned(), path);
        }
    }
    plugins
}

fn valid_name(name: &str) -> bool {
    name.as_bytes()
        .first()
        .map_or(false, u8::is_ascii_lowercase)
        && name
            .bytes()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == b'-')
}

pub fn execute(path: &PathBuf, args: &[OsString]) -> i32 {
    match Command::new(path).args(args).status() {
        Ok(status) => status.code().unwrap_or(126),
        Err(error) => {
            eprintln!("maw: cannot execute {}: {}", path.display(), error);
            126
        }
    }
}
